include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,383 @@
+  childA = parentA % idx + parentB % (1 - idx);
+  // Use traits from parentB for indices where idx is 1 and parentA otherwise.
+  childB = parentA % (1 - idx) + parentA % idx;
+}
+
+//! Perform mutation of the candidates weights with some noise.
+template<typename MatType>
+inline void NSGA2::Mutate(MatType& child)
+{
+  child += (arma::randu<MatType>(child.n_rows, child.n_cols) < mutationProb) %
+      (mutationStrength * arma::randn<MatType>(child.n_rows, child.n_cols));
+}
+
+//! Sort population into Pareto fronts.
+template<typename MatType>
+inline void NSGA2::FastNonDominatedSort(std::vector<std::vector<size_t> >& fronts,
+                                        std::vector<size_t>& ranks,
+                                        std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives)
+{
+  std::map<size_t, size_t> dominationCount;
+  std::map<size_t, std::set<size_t> > dominated;
+
+  // reset and initialize fronts
+  fronts.clear();
+  fronts.push_back(std::vector<size_t>());
+
+  for (size_t p = 0; p < populationSize; p++)
+  {
+    dominated[p] = std::set<size_t>();
+    dominationCount[p] = 0;
+
+    for (size_t q=0; q < populationSize; q++)
+    {
+      if (Dominates<MatType>(calculatedObjectives, p, q))
+      {
+        dominated[p].insert(q);
+      }
+      else if (Dominates<MatType>(calculatedObjectives, q, p))
+      {
+        dominationCount[p] += 1;
+      }
+    }
+
+    if (dominationCount[p] == 0)
+    {
+      ranks[p] = 0;
+      fronts[0].push_back(p);
+    }
+  }
+
+  size_t i = 0;
+
+  while (fronts[i].size() > 0)
+  {
+    std::vector<size_t> nextFront;
+
+    for (size_t p: fronts[i])
+    {
+      for (size_t q: dominated[p])
+      {
+        dominationCount[q]--;
+
+        if (dominationCount[q] == 0)
+        {
+          ranks[q] = i + 1;
+          nextFront.push_back(q);
+        }
+      }
+    }
+
+    i++;
+    fronts.push_back(nextFront);
+  }
+}
+
+//! Check if a candidate Pareto dominates another candidate
+template<typename MatType>
+inline bool NSGA2::Dominates(std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
+                             size_t candidateP,
+                             size_t candidateQ)
+{
+  Info << "NSGA2::NSGA2::Dominates()" << std::endl;
+
+  bool all_better_or_equal = true;
+  bool atleast_one_better = false;
+  size_t n_objectives = calculatedObjectives[0].n_elem;
+
+  for (size_t i = 0; i < n_objectives; i++)
+  {
+    if (calculatedObjectives[candidateP](i) > calculatedObjectives[candidateQ](i))
+    {
+      // p.i is worse than q.i for the i-th objective function
+      all_better_or_equal = false;
+    }
+
+    if (calculatedObjectives[candidateP](i) < calculatedObjectives[candidateQ](i))
+    {
+      // p.i is better than q.i for the i-th objective function
+      atleast_one_better = true;
+    }
+  }
+
+  return all_better_or_equal and atleast_one_better;
```

Comment:
That's interesting, I never knew `and` was a thing in C++, and now I see alternative keywords have existed since C++98! It seems that MSVC does not support this without the `/Za` compiler flag, which is apparently also buggy. Perhaps better to use the normal `&&`?

Reply:
Sure.